nixos/modules/services/network-filesystems/litestream.nix

```diff
@@ -0,0 +1,57 @@
+{ config, lib, options, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.litestream;
+in
+{
+  options.services.litestream = {
+    enable = mkEnableOption "litestream";
+
+    package = mkOption {
+      description = "Package to use.";
+      default = pkgs.litestream;
+      defaultText = "pkgs.litestream";
+      type = types.package;
+    };
+
+    config = mkOption {
+      description = ''
+        See the <link xlink:href="https://litestream.io/reference/config/">documentation</link>.
+      '';
+      type = types.lines;
+    };
+
+    environmentFile = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      example = "/run/secrets/litestream";
+      description = ''
+        Environment file to inject e.g. secrets into the configuration.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+    environment.etc = {
+      "litestream.yml" = {
+        text = cfg.config;
+      };
+    };
+
+    systemd.services.litestream = {
+      description = "Litestream";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "networking.target" ];
+      script = ''
+        exec ${cfg.package}/bin/litestream replicate
+      '';
+      serviceConfig = {
```

Comment:
We don't want to run this as `root`, right?

Reply:
Initially I run this service with a dedicated user but it didn't work with grafana. The default grafana db file permission is `0640` [1]. User running litestream service needs r+w permission to the db file so adding them to grafana group doesn't work. What would be the best approach to make sure that user has required permission? I'm sure grafana isn't the only service creating db files with `0640` permission.

[1] https://github.com/grafana/grafana/blob/dc3e17b/pkg/services/sqlstore/sqlstore.go#L302

Reply:
Ah, yes that makes sense. I'll have to think about this :thinking: 

One option **might** be to add a `user` and `group` option to the `litestream` service and allow users to specify `root` if they plan to work with a number of different services, or maybe run it as the `grafana` user if it will only be replicating the `grafana` database...

I bet @mweinelt would have something smart to say here...

Reply:
The permissions could be changed when litestream is enabled, either by Grafana or in a privileged manner by Litestream. Grafana does not enforce the permissions but will warn out them not matching:

https://github.com/grafana/grafana/blob/dc3e17b/pkg/services/sqlstore/sqlstore.go#L318-L321

If the permissions would be modified to 660 then group permissions would be sufficient, litestream could possibly run with

```nix
preStart = "+" + pkgs.writeShellScript "grant-grafana-db-permissions" ''
  chmod -v 0660 /var/lib/grafana/grafana.db
'';
serviceConfig = { 
  DynamicUser = true;
  User = "litestream";
  SupplementaryGroups = [
    "grafana"
  ];
};

No other great ideas, but indeed services really should not run as root.

Reply:
Thanks for the tip. I've added this bit to the docs to keep it simple. It turns out litestream also needs an execute permission to parent dirs.